### ISSUE TITLE
Option to auto start with system

### ISSUE THREAD

**Issue by eneshecan** [eneshecan/whatsapp-for-linux#86]
Start the app with system boot if the option is enabled in the main menu.

**Comment by berkeyvx** [comment]
I started to work on this issue but there is a few issues. 

If user downloads from github and install via make, `.desktop` file will be created in `/usr/local/share/applications/` But if user installs via snap, `desktop` file will be created in `/snap/whatsapp-for-linux/23/share/applications` (i dont know what is 23 maybe it could changes over the time).

mainWindow.cpp
```cpp
//MainWindow::onStartInTray

    if (visible)
        std::system("cp /usr/share/applications/whatsapp-for-linux.desktop ~/.config/autostart/");
    else
        std::system("rm ~/.config/autostart/whatsapp-for-linux.desktop");
```

If user installs with first option, this code segment is enough to implement it. But for second option, i am not sure if all distros have same directory for snap. We can try to copy and remove from both directories, result will not be a critical if command fails.

I need a little help about what is 23 and Is Snap in the same directory in all distributions

**Comment by eneshecan** [comment]
Thanks, @berkeyvx! Yes, snap needs to be adapted as well but I have no experience with it. I'll update here if I find out how can we do this in snap environment.

**Comment by berkeyvx** [comment]
We can create our .desktop file while application starts. Whenever button toggles, we remove or create .desktop file in directory ~/.config/autostart. With this approach we do not need to think about snap. What do you think about this ? @eneshecan

**Comment by eneshecan** [comment]
> We can create our .desktop file while application starts. Whenever button toggles, we remove or create .desktop file in directory ~/.config/autostart. With this approach we do not need to think about snap. What do you think about this ? @eneshecan

This is like a workaround I would avaoid it. Instead we can do it like that;

- Get the applications folder string by using some environment variable which also should work in snap
- Append our desktop file name to the end
- The rest should be the same as your code pasted above

What do you think?

**Comment by berkeyvx** [comment]
It is better i guess. Did some research about that. We can do this in two different ways.

* Using unix header file named unistd.h which contains `getcwd` function
* C++17 `filesystem` library using `current_path` function

2nd option may require additional compiler/linker options. According to cppreference:
`GNU implementation prior to 9.1 requires linking with -lstdc++fs and LLVM implementation prior to LLVM 9.0 requires linking with -lc++fs.`

Which one you prefer?

**Comment by eneshecan** [comment]
Try this one since it's glib
https://developer.gnome.org/glib/stable/glib-Miscellaneous-Utility-Functions.html#g-get-user-data-dir

**Comment by berkeyvx** [comment]
both of my suggestion and your suggestion does not work. 

* `fs::current_path` returns `/home/berkeyvx`
* `g-get-user-data-dir` returns `/home/berkeyvx/.local/share`

`make install` command installs under `/usr/local/share` directory. 
We can move CMake install to .local/share directory.

**Comment by eneshecan** [comment]
No, if you pass `-DCMAKE_INSTALL_PREFIX=/usr` it will install into `/usr` dir, otherwise it installs into `HOME/.local/`. So, I guess `g-get-user-data-dir` is the way to go, the only thing you have to do is appending `/applications/` to the end.

**Comment by eneshecan** [comment]
I'll take a look...

**Comment by eneshecan** [comment]
Found this for snap; https://forum.snapcraft.io/t/how-to-add-autostart-to-the-app/8866 but couldn't find anything solid for the normal case yet...

**Comment by eneshecan** [comment]
Okay, I think the following is feasible; since the desktop entry file in one of the folders `usr/share/applications` or `usr/local/share/applications` we can introduce a function that checks the location of our desktop entry file by trying those two and returns the full path. Then your algorithm can take care the rest. But please try to use Glib functions while copying/removing the file. See https://developer.gnome.org/gio/stable/GFile.html#g-file-copy and https://developer.gnome.org/gio/stable/GFile.html#g-file-delete. Also, maybe we can try use the `async` version of them to not block the main gui thread but I don't think it's needed since the file is very small.
